include/vara/Feature/FeatureModel.h

```diff
@@ -1,96 +1,206 @@
+  end() const {
+    return OrderedFeatures.end();
+  }
+
+  llvm::iterator_range<OrderedFeatureVector::ordered_feature_iterator>
+  features() {
+    return llvm::make_range(begin(), end());
+  }
+  [[nodiscard]] llvm::iterator_range<
+      OrderedFeatureVector::const_ordered_feature_iterator>
+  features() const {
+    return llvm::make_range(begin(), end());
+  }
+
+  //===--------------------------------------------------------------------===//
+  // Utility
 
   void view() { ViewGraph(this, "FeatureModel-" + this->getName()); }
 
-  struct FeatureModelIter : std::iterator<std::forward_iterator_tag, Feature &,
-                                          ptrdiff_t, Feature *, Feature &> {
-    FeatureMapTy::iterator It;
+  Feature *getFeature(llvm::StringRef F) { return Features[F].get(); }
 
-    explicit FeatureModelIter(FeatureMapTy::iterator It) : It(It) {}
+  LLVM_DUMP_METHOD
+  void dump() const;
 
-    Feature *operator*() const { return It->second.get(); }
+protected:
+  string Name;
+  fs::path Path;
+  FeatureMapTy Features;
+  Feature *Root;
 
-    FeatureModelIter &operator++() {
-      It++;
-      return *this;
-    }
+private:
+  OrderedFeatureTy OrderedFeatures;
+};
 
-    bool operator==(const FeatureModelIter &Other) const {
-      return It == Other.It;
-    }
+//===----------------------------------------------------------------------===//
+//                     Builder for FeatureModel
+//===----------------------------------------------------------------------===//
 
-    bool operator!=(const FeatureModelIter &Other) const {
-      return not operator==(Other);
-    }
-  };
+/// \brief Builder for \a FeatureModel which can be used while parsing.
+class FeatureModelBuilder : private FeatureModel {
+public:
+  FeatureModelBuilder() : FeatureModel("", "", {}, nullptr){};
+
+  void init() {
+    Name = "";
+    Path = "";
+    Root = nullptr;
+    Features.clear();
+    Constraints.clear();
+    Parents.clear();
+    Children.clear();
+    Excludes.clear();
+  }
 
-  FeatureModelIter begin() { return FeatureModelIter(Features.begin()); }
+  /// Try to add \a BinaryFeature.
+  bool addFeature(const std::string &Key, bool Opt = false,
+                  std::optional<FeatureSourceRange> Loc = std::nullopt) {
+    return Features
+        .try_emplace(Key,
+                     std::make_unique<BinaryFeature>(Key, Opt, std::move(Loc)))
+        .second;
+  }
 
-  FeatureModelIter end() { return FeatureModelIter(Features.end()); }
+  /// Try to add \a NumericFeature.
+  bool addFeature(const std::string &Key,
+                  const NumericFeature::ValuesVariantType &Values,
+                  bool Opt = false,
+                  std::optional<FeatureSourceRange> Loc = std::nullopt) {
+    return Features
+        .try_emplace(Key, std::make_unique<NumericFeature>(Key, Values, Opt,
+                                                           std::move(Loc)))
+        .second;
+  }
```

Comment:
We should merge these functions into a factory method and rename them to `make_feature` or `create_feature`, to not be confused with the `addFeature(Feature &)` method, which has a different behavior (Add a feature by copying it vs creating a new feature). Furthermore, `Key` is not a really good name here, it is the key for the map but from the API users side it is the feature name, and we should not confuse the user by naming things based on our implementation details.

Regarding the factory method:
```
    /// Try to create and add a new \a Feature to the \a FeatureModel.
  ///
  /// \param[in] FeatureName name of the \a Feature
  /// \param[in] FurtherArgs further arguments that should be passed to the
  ///                        \a Feature constructor
  ///
  /// \returns true, if the feature could be inserted into the \a FeatureModel
  template <
      typename FeatureTy, typename... Args,
      typename std::enable_if_t<std::is_base_of_v<Feature, FeatureTy>, int>>
  bool makeFeature(const std::string &FeatureName, Args... FurtherArgs) {
    return Features
        .try_emplace(FeatureName, std::make_unique<FeatureTy>(
                                      FeatureName, std::forward(FurtherArgs)...))
        .second;
  }
```
This allows users to create any kind of feature as long as it is derived from `Feature` and the passed args match one of the specific Features ctors.